### PR TITLE
UI: Add Trovo integration

### PR DESCRIPTION
**Pull request by tlivegaming**
### Description
Add Trovo integration so that users can connect OBS to Trovo account

### Motivation and Context
Add Trovo integration. https://trovo.live.

### How Has This Been Tested?
Server tested and working properly.

### Types of changes
New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Review by RytoEX — COMMENTED** [review]
I have some questions about this PR.

File: UI/auth-trovo.cpp [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+	key_ = json["stream_key"].string_value();
+
+	std::string channelUrl = json["channel_url"].string_value();
+	int index = channelUrl.find_last_of("/") + 1;
+	if (index > 0 && index < (int)channelUrl.size()) {
+		name = channelUrl.substr(index);
+	}
+
+	/** Trovo host url already has "live/", so ingore "live/" in stream_key */
```
[the remark this answers is not in the document]
Comment (jp9000) [reply UI/auth-trovo.cpp:91]: The comment style doesn't really matter that much.

File: UI/auth-trovo.cpp [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+#include "window-basic-main.hpp"
+#include "remote-text.hpp"
+#include "ui-config.h"
+#include "obf.h"
+
+using namespace json11;
+
+/* ------------------------------------------------------------------------- */
+#define TROVO_AUTH_URL \
+	"https://open.trovo.live/page/login.html?client_id=6W4TpDJnE95U68zWJHZLDfn4FA5wf7W4&response_type=code&scope=channel_details_self+channel_update_self+user_details_self&redirect_uri=https%3A%2F%2Fobsproject.com%2Fapp-auth%2Ftrovo-auth.php&state=Nlc0VHBESm5FOTVVNjh6V0pIWkxEZm40RkE1d2Y3VzQ"
```
[the remark this answers is not in the document]
Comment (tlivegaming) [reply UI/auth-trovo.cpp:22]: The client ID and hash can be passed in via CMake at build time by setting TROVO_CLIENTID and TROVO_HASH params.
I have ever consulted OBS team in discord about the AUTH_URL. In my understanding, you will create fresh addresses with redirects and replace TROVO_AUTH_URL when merging(just like `#define TWITCH_AUTH_URL "https://obsproject.com/app-auth/twitch?action=redirect"`) . So I just hardcode the TROVO_AUTH_URL here. Would you do that? Or should I get the client ID from CMake and random the state parameter and then generate the TROVO_AUTH_URL with them?
image

File: UI/auth-trovo.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tlivegaming) [reply UI/auth-trovo.cpp]: We have fixed it.

**Comment by tlivegaming** [comment]
- 

> I have some questions about this PR.

Now we have resolved all issues and conflicts

File: UI/auth-trovo.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (RytoEX) [reply UI/auth-trovo.cpp]: You never answered this question.

[the remark this answers is not in the document]
Comment (tlivegaming) [reply UI/auth-trovo.cpp]: I think Jim will generate this random state and add the command below at build time
```
-DTROVO_HASH=0 -DTROVO_CLIENTID=6W4***5wf7W4
```
Should I replace TROVO_AUTH_URL, "https://open.trovo.live/page/login.html?client_id=6W4TpDJnE95U..." with a link like "https://obsproject.com/app-auth/trovo? action=redirect" now?


image

**Comment by tlivegaming** [comment]
> @RytoEX I think Jim will generate this random state and add the command below at build time
> 
> ```
> -DTROVO_HASH=0 -DTROVO_CLIENTID=6W4***5wf7W4
> ```
> 
> Should I replace TROVO_AUTH_URL, "https://open.trovo.live/page/login.html?client_id=6W4TpDJnE95U..." with a link like "https://obsproject.com/app-auth/trovo? action=redirect" now?
> 
> image

File: UI/auth-trovo.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tlivegaming) [reply UI/auth-trovo.cpp]: I'm waiting for a reply and make some adjustments.

**Comment by derrod** [comment]
Apologies for only getting back to you now, since your message is all quoted it probably got lost in emails.

We have changed how our OAuth URLs work, we now use a define for the base, and you only need to specify the path. For example in the Twitch implementation: 
https://github.com/obsproject/obs-studio/blob/master/UI/auth-twitch.cpp#L25-L26
So for Trovo you'd probably want use something along the lines of
```cpp
#define TROVO_AUTH_URL OAUTH_BASE_URL "v1/trovo/redirect"
#define TROVO_TOKEN_URL OAUTH_BASE_URL "v1/trovo/token"
```

Our server-side component is now also open source and can be found here: https://github.com/obsproject/obs-oauth-cf
Updating it for Trovo support should be trivial.

**Comment by derrod** [comment]
Closing due to lack of response.

**Pull request closed without merging**